gossipsub.go

```diff
@@ -238,17 +368,49 @@ func (gs *GossipSubRouter) handleIWant(p peer.ID, ctl *pb.ControlMessage) []*pb.
 
 func (gs *GossipSubRouter) handleGraft(p peer.ID, ctl *pb.ControlMessage) []*pb.ControlPrune {
 	var prune []string
+
+	doPX := true
+	score := gs.score.Score(p)
+	now := time.Now()
+
 	for _, graft := range ctl.GetGraft() {
 		topic := graft.GetTopicID()
 		peers, ok := gs.mesh[topic]
 		if !ok {
+			// don't do PX when there is an unknown topic to avoid leaking our peers
+			doPX = false
+			// spam harndening: ignore GRAFTs for unknown topics
+			continue
+		}
+
+		// check the score
+		if score < 0 {
+			// we don't GRAFT peers with negative score
+			log.Debugf("GRAFT: ignoring peer %s with negative score [score = %f, topic = %s]", p, score, topic)
+			// we do send them PRUNE however, because it's a matter of protocol correctness
```

Comment:
They will be backed off and if they try repeatedly immediately after they will never get in because they will keep extending their backoff period.
Otherwise they will have to wait at least 1min to try again.

I think we should still emit the `PRUNE` message here, because it's a matter of protocol correctness.

Consider an honest node that happens to have a negative score (eg because of staked participation and them not having put in the stake yet, or because it was in the mesh and got pruned because it was not delivering enough messages). If we don't emit a PRUNE here they will think they are in the mesh while they are not and fail to build a mesh.